include/seqan/bam_io/bam_index_bai.h

```diff
@@ -154,6 +154,151 @@ class BamIndex<Bai>
+
+    list.clear();
+
+    unsigned k;
+    list.push_back(0);
+    for (k =    1 + (beg>>26); k <=    1 + (end>>26); ++k) list.push_back(k);
+    for (k =    9 + (beg>>23); k <=    9 + (end>>23); ++k) list.push_back(k);
+    for (k =   73 + (beg>>20); k <=   73 + (end>>20); ++k) list.push_back(k);
+    for (k =  585 + (beg>>17); k <=  585 + (end>>17); ++k) list.push_back(k);
+    for (k = 4681 + (beg>>14); k <= 4681 + (end>>14); ++k) list.push_back(k);
+}
+
+template <typename TContainer, typename TSpec>
+inline
+SEQAN_FUNC_ENABLE_IF(And<IsSameType<typename Value<TContainer>::Type, BamAlignmentRecord>,
+                     Not<IsSameType<typename Value<TContainer>::Type, TContainer>>>, bool)
+viewRecords(TContainer & resultContainer,
+            FormattedFile<Bam, Input, TSpec> & bamFile,
+            BamIndex<Bai> const & bamIndex,
+            int32_t refId,
+            int32_t regionStart,
+            int32_t regionEnd)
+{
+    typedef typename std::map<uint32_t, BaiBamIndexBinData_>::const_iterator TMapIter;
+    typedef String<Pair<uint64_t, uint64_t>> TChunk;
+    typedef Iterator<TChunk const, Rooted>::Type TChunkIter;
+
+    // Sanity checks
+    // -------------------------------------------------------------------------
+    if (!isEqual(format(bamFile), Bam()))
+        return false;
+
+    if (refId < 0)
+        return false;  // Cannot seek to invalid reference.
+
+    if (static_cast<unsigned>(refId) >= length(bamIndex._binIndices))
+        return false;  // Cannot seek to invalid reference.
+
+    if (regionStart > regionEnd)
+        return false; // invalid region
+
+    if (regionEnd > 536870912)
+        return false; // BAM format only allows reference of size < 2^29
+
+    // Adjust region
+    // -------------------------------------------------------------------------
+    // We want to mimic samtools view. Since SAM/BAM files are one based we expect
+    // the given region to be also one-based (To be consistent with samtools).
+    // Since SeqAn works 0-based we need to substract 1.
+    SEQAN_ASSERT_GEQ_MSG(regionStart, 1, "Expect 1-based region interval which excludes 0 as a start pos.");
+    --regionStart;
+    --regionEnd;
+
+    // Retrieve the candidate bin identifiers for [regionStart, regionEnd).
+    // -------------------------------------------------------------------------
+    std::vector<uint16_t> candidateBins;
+    _baiReg2bins(candidateBins, regionStart, regionEnd);
+
+    // Retrieve the smallest required offset from the linear index.
+    // -------------------------------------------------------------------------
+    uint64_t linearMinOffset;
+    if (empty(bamIndex._linearIndices[refId]))
+    {
+        linearMinOffset = 0; // if no linear index is given, value 0 never limits chunk seeking
+    }
+    else
+    {
+        unsigned windowIdx = regionStart >> 14; // Linear index consists of 16kb windows.
+        linearMinOffset = bamIndex._linearIndices[refId][windowIdx];
```

Comment:
Hmmm. This is strange. Since `regionStart` is within the length of the sequence (I suppose you didn't give abnormal bounds) I thought this could never happen.. because the index stores an offsett for ever 16kb window of the reference.. and if we do not store empty windows the access to `linearIndice` is not trivial.

Reply:
It seems it happens when I specify a region that is in a index window/bin that is after the last read's window. For example, if I have a BAM file with one 100 bp read at position 60000 then querying [20000-29000] works fine (no reads found) but querying [80000-89000] throws an assertion failure. I think we just need a little check for this case.

Reply:
What is the length of your reference, specified in the BAM header, for this example?

Reply:
`LN:100000` but it does not seem to have any effect if I change it. `LN:1` and `LN:10000000` also gives my identical results.

Reply:
Thanks for trying it. Then I guess the linear index ends either when the reference length is reached or there are no more records(/end of file is reached). I'll add a check.

Reply:
Thanks, I will give it a try when it is ready. And sorry that I wasn't very clear before, but what I meant was that the linear index length is not depended on the reference length (`LN`) at all. The assertion failure only happens when you specify a region that is behind the last record on that contig/chromosome (does not need to be the end of the file). If the last record is much further than the reference length the linear index also reaches as far as the last record.

For example, I was still able to find the read at position 60000 even when I had `LN:1` in the BAM header. In that case the linear index reaches much further than the specified reference length.